File: hw1/shingling.py
from collections import Counter
import hashlib

class Shingling:
    def __init__(self, k=10):
        self.k = k
        self.shingles = None
        self.hshr = hashlib.md5()

    def compute_shingles(self, filename):

        with open(filename, 'r') as f:
            curr_shingl = f.read(self.k)
            self.shingles = set([self._hasher(curr_shingl)])

            while True:
                c = f.read(1)
                if not c:
                    break
                curr_shingl = curr_shingl[1:] + c
                self.shingles.add(self._hasher(curr_shingl))
        return self.shingles

    def _hasher(self, string, size=4):
        self.hshr.update(string)
        return int(self.hshr.hexdigest()[-size:], 16)

File: hw1/src/Shingling.java
import java.io.FileReader;
import java.io.Reader;
import java.io.BufferedReader;
import java.io.IOException;

import java.util.ArrayList;

class Shingling {
    private int k;

    Shingling(int k) {
        this.k = k;
    }

    ArrayList<String> shingles(String filename) {
        Reader r = null;

        try {
            r = new BufferedReader(new FileReader(filename));
            String str = "";
            int rd;

            while((rd = r.read()) != -1) {
                char c = (char)rd;
                str += c;
                System.out.println(str);
            }

        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                if (r != null) r.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }

        return null;
    }
}
